src/ott/neural/solvers/expectile_neural_dual.py

```diff
@@ -0,0 +1,655 @@
+          batch,
+      )
+
+      return (
+          state_f.apply_gradients(grads=grads_f),
+          state_g.apply_gradients(grads=grads_g), loss, loss_f, loss_g, w_dist
+      )
+
+    return step_fn
+
+  def _get_valid_step(
+      self
+  ) -> Callable[
+      [PotentialTrainState, PotentialTrainState, Dict[str, jnp.ndarray]], Tuple[
+          jnp.ndarray, jnp.ndarray, jnp.ndarray]]:
+    loss_fn = self._euclidean_loss_fn if self.use_dot_product else self._loss_fn
+
+    @jax.jit
+    def step_fn(state_f, state_g, batch):
+      grad_fn = jax.value_and_grad(loss_fn, argnums=[0, 1], has_aux=True)
+      (loss, (loss_f, loss_g, w_dist)), _ = grad_fn(
+          state_f.params,
+          state_g.params,
+          state_f.potential_gradient_fn,
+          state_g.potential_value_fn,
+          batch,
+      )
+
+      return loss_f, loss_g, w_dist
+
+    return step_fn
+
+  def _loss_fn(
+      self, params_f: frozen_dict.FrozenDict[str, jnp.ndarray],
+      params_g: frozen_dict.FrozenDict[str, jnp.ndarray],
+      gradient_f: Callable[[frozen_dict.FrozenDict[str, jnp.ndarray]],
+                           PotentialGradientFn_t],
+      g_value: Callable[[frozen_dict.FrozenDict[str, jnp.ndarray]],
+                        PotentialValueFn_t], batch: Dict[str, jnp.ndarray]
+  ) -> Tuple[jnp.ndarray, Tuple[jnp.ndarray, jnp.ndarray, jnp.ndarray]]:
+
+    if self.is_bidirectional:
+      grad_h_inv = jax.vmap(jax.grad(self.cost_fn.h_legendre))
+      transport = lambda grad_f, source: source - grad_h_inv(grad_f(source))
+    else:
+      transport = lambda grad_f, source: grad_f(source)
+
+    def g_c_transform(x, y):
+      return batch_cost(x,y) - \
+        jax.vmap(g_value(jax.lax.stop_gradient(params_g)))(y)
+
+    source, target = batch["source"], batch["target"]
+
+    f_grad_partial = jax.vmap(gradient_f(params_f))
+    g_value_partial = jax.vmap(g_value(params_g))
+    batch_cost = jax.vmap(self.cost_fn)
+
+    target_hat = transport(f_grad_partial, source)
+    target_hat_detach = jax.lax.stop_gradient(target_hat)
+
+    g_target = g_value_partial(target)
+    g_star_source = batch_cost(source, target_hat_detach
+                              ) - g_value_partial(target_hat_detach)
+
+    diff_1 = jax.lax.stop_gradient(
+        -batch_cost(source, target) + g_c_transform(source, target_hat_detach)
+    ) + g_target
+    reg_loss_1 = expectile_loss(diff_1, diff_1, self.expectile).mean()
+
+    diff_2 = jax.lax.stop_gradient(
+        -g_c_transform(source, target)
+    ) + g_star_source
+    reg_loss_2 = expectile_loss(diff_2, diff_2, self.expectile).mean()
+
+    reg_loss = (reg_loss_1 + reg_loss_2) * self.expectile_loss_coef
+    dual_loss = -(g_target.mean() + g_star_source.mean())
+    amor_loss = g_c_transform(source, target_hat).mean()
+
+    loss = reg_loss + dual_loss + amor_loss
+
+    w_dist = (g_target.mean() + g_star_source.mean())
+
+    return loss, (dual_loss, amor_loss, w_dist)
+
+  def _euclidean_loss_fn(
```

Comment:
Possibly a naive question, but why is `DotCost` treated differently here - shouldn't it be equivalent (at least in theory the `def _loss_fn` with the `DotCost`? Or is this more numerically stable impl.?